File: Cpp/Highlighters.cpp
#include "Highlighters.h"

#include <QFileInfo>
#include <QFont>

namespace {

QTextCharFormat fmt(const QString &color, bool bold = false, bool italic = false)
{
    QTextCharFormat f;
    f.setForeground(QColor(color));
    if (bold)
        f.setFontWeight(QFont::Bold);
    if (italic)
        f.setFontItalic(true);
    return f;
}

// Python 3's keyword.kwlist.
const QStringList kPythonKeywords = {
    "False",  "None",     "True",  "and",    "as",       "assert", "async",  "await",
    "break",  "class",    "continue", "def",  "del",      "elif",   "else",   "except",
    "finally", "for",     "from",  "global", "if",       "import", "in",     "is",
    "lambda", "nonlocal", "not",   "or",     "pass",     "raise",  "return", "try",
    "while",  "with",     "yield",
};

const QStringList kPythonBuiltins = {
    "self", "cls", "None", "True", "False", "print", "len", "range",
    "int", "str", "float", "list", "dict", "set", "tuple",
};

} // namespace

PythonHighlighter::PythonHighlighter(QTextDocument *document)
    : QSyntaxHighlighter(document)
{
    m_rules.append({QRegularExpression("\\b(" + kPythonKeywords.join('|') + ")\\b"), fmt("#af00db", true)});
    m_rules.append({QRegularExpression("\\b(" + kPythonBuiltins.join('|') + ")\\b"), fmt("#267f99")});
    m_rules.append({QRegularExpression("\\bdef\\s+(\\w+)"), fmt("#795e26")});
    m_rules.append({QRegularExpression("\\bclass\\s+(\\w+)"), fmt("#267f99", true)});
    m_rules.append({QRegularExpression("@\\w+"), fmt("#795e26", false, true)});
    m_rules.append({QRegularExpression("\\b[0-9]+\\.?[0-9]*\\b"), fmt("#098658")});
    m_rules.append({QRegularExpression("'[^'\\\\]*(\\\\.[^'\\\\]*)*'"), fmt("#a31515")});
    m_rules.append({QRegularExpression("\"[^\"\\\\]*(\\\\.[^\"\\\\]*)*\""), fmt("#a31515")});
    m_rules.append({QRegularExpression("#[^\\n]*"), fmt("#008000", false, true)});

    m_tripleFormat = fmt("#a31515");
    m_tripleSingle = QRegularExpression("'''");
    m_tripleDouble = QRegularExpression("\"\"\"");
}

void PythonHighlighter::highlightBlock(const QString &text)
{
    for (const auto &rule : m_rules) {
        auto it = rule.first.globalMatch(text);
        while (it.hasNext()) {
            const QRegularExpressionMatch m = it.next();
            setFormat(m.capturedStart(), m.capturedLength(), rule.second);
        }
    }

    highlightTriple(text, m_tripleDouble, 1);
    highlightTriple(text, m_tripleSingle, 2);
}

void PythonHighlighter::highlightTriple(const QString &text, const QRegularExpression &pattern, int stateId)
{
    int start = 0;
    if (previousBlockState() == stateId) {
        const QRegularExpressionMatch match = pattern.match(text);
        const int end = match.hasMatch() ? match.capturedStart() : -1;
        if (end == -1) {
            setFormat(0, text.length(), m_tripleFormat);
            setCurrentBlockState(stateId);
            return;
        }
        setFormat(0, end + 3, m_tripleFormat);
        start = end + 3;
    }

    while (true) {
        const QRegularExpressionMatch match = pattern.match(text, start);
        if (!match.hasMatch())
            break;
        const int openStart = match.capturedStart();
        const QRegularExpressionMatch match2 = pattern.match(text, openStart + 3);
        if (!match2.hasMatch()) {
            setFormat(openStart, text.length() - openStart, m_tripleFormat);
            setCurrentBlockState(stateId);
            return;
        }
        const int closeEnd = match2.capturedEnd();
        setFormat(openStart, closeEnd - openStart, m_tripleFormat);
        start = closeEnd;
    }
}

JsonHighlighter::JsonHighlighter(QTextDocument *document)
    : QSyntaxHighlighter(document)
{
    m_rules.append({QRegularExpression("\"[^\"\\\\]*(\\\\.[^\"\\\\]*)*\"\\s*(?=:)"), fmt("#0451a5")});
    m_rules.append({QRegularExpression("(?<=:)\\s*\"[^\"\\\\]*(\\\\.[^\"\\\\]*)*\""), fmt("#a31515")});
    m_rules.append({QRegularExpression("\\b-?[0-9]+\\.?[0-9]*([eE][+-]?[0-9]+)?\\b"), fmt("#098658")});
    m_rules.append({QRegularExpression("\\b(true|false|null)\\b"), fmt("#0000ff", true)});
}

void JsonHighlighter::highlightBlock(const QString &text)
{
    for (const auto &rule : m_rules) {
        auto it = rule.first.globalMatch(text);
        while (it.hasNext()) {
            const QRegularExpressionMatch m = it.next();
            setFormat(m.capturedStart(), m.capturedLength(), rule.second);
        }
    }
}

MarkdownHighlighter::MarkdownHighlighter(QTextDocument *document)
    : QSyntaxHighlighter(document)
{
    m_rules.append({QRegularExpression("^#{1,6}\\s.*"), fmt("#0451a5", true)});
    m_rules.append({QRegularExpression("\\*\\*[^*]+\\*\\*"), fmt("#24292f", true)});
    m_rules.append({QRegularExpression("(?<!\\*)\\*[^*]+\\*(?!\\*)"), fmt("#24292f", false, true)});
    m_rules.append({QRegularExpression("`[^`]+`"), fmt("#a31515")});
    m_rules.append({QRegularExpression("\\[[^\\]]*\\]\\([^)]*\\)"), fmt("#0969da")});
    m_rules.append({QRegularExpression("^\\s*[-*+]\\s"), fmt("#af00db")});
    m_rules.append({QRegularExpression("^>.*"), fmt("#6a737d", false, true)});
}

void MarkdownHighlighter::highlightBlock(const QString &text)
{
    for (const auto &rule : m_rules) {
        auto it = rule.first.globalMatch(text);
        while (it.hasNext()) {
            const QRegularExpressionMatch m = it.next();
            setFormat(m.capturedStart(), m.capturedLength(), rule.second);
        }
    }
}

HighlighterKind highlighterKindFor(const QString &filePath)
{
    if (filePath.isEmpty())
        return HighlighterKind::None;
    const QString ext = QFileInfo(filePath).suffix().toLower();
    if (ext == "py" || ext == "pyw")
        return HighlighterKind::Python;
    if (ext == "json")
        return HighlighterKind::Json;
    if (ext == "md" || ext == "markdown")
        return HighlighterKind::Markdown;
    return HighlighterKind::None;
}

QSyntaxHighlighter *createHighlighter(HighlighterKind kind, QTextDocument *document)
{
    switch (kind) {
    case HighlighterKind::Python:
        return new PythonHighlighter(document);
    case HighlighterKind::Json:
        return new JsonHighlighter(document);
    case HighlighterKind::Markdown:
        return new MarkdownHighlighter(document);
    case HighlighterKind::None:
    default:
        return nullptr;
    }
}

File: Python/highlighters.py
import keyword
import os

from PyQt6.QtCore import QRegularExpression
from PyQt6.QtGui import QColor, QFont, QSyntaxHighlighter, QTextCharFormat


def _fmt(color, bold=False, italic=False):
    f = QTextCharFormat()
    f.setForeground(QColor(color))
    if bold:
        f.setFontWeight(QFont.Weight.Bold)
    if italic:
        f.setFontItalic(True)
    return f


class PythonHighlighter(QSyntaxHighlighter):
    def __init__(self, document):
        super().__init__(document)
        self.rules = []

        keyword_fmt = _fmt("#af00db", bold=True)
        pattern = r"\b(" + "|".join(keyword.kwlist) + r")\b"
        self.rules.append((QRegularExpression(pattern), keyword_fmt))

        builtin_fmt = _fmt("#267f99")
        builtins = ["self", "cls", "None", "True", "False", "print", "len", "range", "int", "str", "float", "list", "dict", "set", "tuple"]
        pattern = r"\b(" + "|".join(builtins) + r")\b"
        self.rules.append((QRegularExpression(pattern), builtin_fmt))

        self.rules.append((QRegularExpression(r"\bdef\s+(\w+)"), _fmt("#795e26")))
        self.rules.append((QRegularExpression(r"\bclass\s+(\w+)"), _fmt("#267f99", bold=True)))
        self.rules.append((QRegularExpression(r"@\w+"), _fmt("#795e26", italic=True)))
        self.rules.append((QRegularExpression(r"\b[0-9]+\.?[0-9]*\b"), _fmt("#098658")))
        self.rules.append((QRegularExpression(r"'[^'\\]*(\\.[^'\\]*)*'"), _fmt("#a31515")))
        self.rules.append((QRegularExpression(r'"[^"\\]*(\\.[^"\\]*)*"'), _fmt("#a31515")))
        self.rules.append((QRegularExpression(r"#[^\n]*"), _fmt("#008000", italic=True)))

        self.triple_fmt = _fmt("#a31515")
        self.triple_single = QRegularExpression(r"'''")
        self.triple_double = QRegularExpression(r'"""')

    def highlightBlock(self, text):
        for pattern, fmt in self.rules:
            it = pattern.globalMatch(text)
            while it.hasNext():
                m = it.next()
                self.setFormat(m.capturedStart(), m.capturedLength(), fmt)

        self._highlight_triple(text, self.triple_double, 1)
        self._highlight_triple(text, self.triple_single, 2)

    def _highlight_triple(self, text, pattern, state_id):
        start = 0
        if self.previousBlockState() == state_id:
            match = pattern.match(text)
            end = match.capturedStart() if match.hasMatch() else -1
            if end == -1:
                self.setFormat(0, len(text), self.triple_fmt)
                self.setCurrentBlockState(state_id)
                return
            self.setFormat(0, end + 3, self.triple_fmt)
            start = end + 3

        while True:
            match = pattern.match(text, start)
            if not match.hasMatch():
                break
            open_start = match.capturedStart()
            match2 = pattern.match(text, open_start + 3)
            if not match2.hasMatch():
                self.setFormat(open_start, len(text) - open_start, self.triple_fmt)
                self.setCurrentBlockState(state_id)
                return
            close_end = match2.capturedEnd()
            self.setFormat(open_start, close_end - open_start, self.triple_fmt)
            start = close_end


class JsonHighlighter(QSyntaxHighlighter):
    def __init__(self, document):
        super().__init__(document)
        self.rules = [
            (QRegularExpression(r'"[^"\\]*(\\.[^"\\]*)*"\s*(?=:)'), _fmt("#0451a5")),
            (QRegularExpression(r'(?<=:)\s*"[^"\\]*(\\.[^"\\]*)*"'), _fmt("#a31515")),
            (QRegularExpression(r"\b-?[0-9]+\.?[0-9]*([eE][+-]?[0-9]+)?\b"), _fmt("#098658")),
            (QRegularExpression(r"\b(true|false|null)\b"), _fmt("#0000ff", bold=True)),
        ]

    def highlightBlock(self, text):
        for pattern, fmt in self.rules:
            it = pattern.globalMatch(text)
            while it.hasNext():
                m = it.next()
                self.setFormat(m.capturedStart(), m.capturedLength(), fmt)


class MarkdownHighlighter(QSyntaxHighlighter):
    def __init__(self, document):
        super().__init__(document)
        self.rules = [
            (QRegularExpression(r"^#{1,6}\s.*"), _fmt("#0451a5", bold=True)),
            (QRegularExpression(r"\*\*[^*]+\*\*"), _fmt("#24292f", bold=True)),
            (QRegularExpression(r"(?<!\*)\*[^*]+\*(?!\*)"), _fmt("#24292f", italic=True)),
            (QRegularExpression(r"`[^`]+`"), _fmt("#a31515")),
            (QRegularExpression(r"\[[^\]]*\]\([^)]*\)"), _fmt("#0969da")),
            (QRegularExpression(r"^\s*[-*+]\s"), _fmt("#af00db")),
            (QRegularExpression(r"^>.*"), _fmt("#6a737d", italic=True)),
        ]

    def highlightBlock(self, text):
        for pattern, fmt in self.rules:
            it = pattern.globalMatch(text)
            while it.hasNext():
                m = it.next()
                self.setFormat(m.capturedStart(), m.capturedLength(), fmt)


EXTENSION_MAP = {
    ".py": PythonHighlighter,
    ".pyw": PythonHighlighter,
    ".json": JsonHighlighter,
    ".md": MarkdownHighlighter,
    ".markdown": MarkdownHighlighter,
}


def highlighter_class_for(file_path):
    if not file_path:
        return None
    ext = os.path.splitext(file_path)[1].lower()
    return EXTENSION_MAP.get(ext)
